Applications/Spire/Source/Ui/OverlayPanel.cpp

```diff
@@ -171,25 +159,35 @@ void OverlayPanel::position() {
     auto parent_geometry = parentWidget()->rect();
     auto parent_bottom_left = parentWidget()->mapToGlobal(
       parent_geometry.bottomLeft());
-    auto screen_geometry =
-      get_current_screen(parent_bottom_left)->availableGeometry();
+    auto screen_geometry = parentWidget()->screen()->availableGeometry();
+    auto panel_size = [=] {
+      if(layout()->contentsMargins() == DROP_SHADOW_MARGINS()) {
+        return size() - QSize(0, DROP_SHADOW_HEIGHT());
+      }
+      return size();
+    }();
     auto x = parent_bottom_left.x() - DROP_SHADOW_WIDTH();
-    if(x < 0) {
-      x = 0;
-    } else if(x + width() > screen_geometry.right()) {
-      x = screen_geometry.right() - width();
-    }
-    if((parent_bottom_left.y() + height()) > screen_geometry.bottom()) {
-      layout()->setContentsMargins(DROP_SHADOW_WIDTH(), DROP_SHADOW_HEIGHT(),
-        DROP_SHADOW_WIDTH(), 0);
-      update();
-      move({x, parent_bottom_left.y() - parent_geometry.height() - height() +
-        scale_height(1)});
-    } else {
-      layout()->setContentsMargins(DROP_SHADOW_WIDTH(), 0,
-        DROP_SHADOW_WIDTH(), DROP_SHADOW_HEIGHT());
-      update();
-      move({x, parent_bottom_left.y() + scale_height(1)});
+    if(x < screen_geometry.left()) {
+      x = screen_geometry.left() - DROP_SHADOW_WIDTH();
+    } else if(x + panel_size.width() > screen_geometry.right()) {
+      x = screen_geometry.right() - panel_size.width() + DROP_SHADOW_WIDTH();
     }
+    auto rect = [=] () -> QRect {
+      if((parent_bottom_left.y() + panel_size.height()) >
+          screen_geometry.bottom()) {
+        auto margins = QMargins(DROP_SHADOW_WIDTH(), DROP_SHADOW_HEIGHT(),
+          DROP_SHADOW_WIDTH(), 0);
+        layout()->setContentsMargins(margins);
+        return {QPoint(x, parent_bottom_left.y() - parent_geometry.height() -
+          panel_size.height() + 1), panel_size};
+      } else {
+        auto margins = QMargins(DROP_SHADOW_WIDTH(), 0, DROP_SHADOW_WIDTH(),
+          DROP_SHADOW_HEIGHT());
+        layout()->setContentsMargins(margins);
+        return {QPoint(x, parent_bottom_left.y() + 1), panel_size};
+      }
+    }();
+    setGeometry(rect);
+    update();
```

Comment:
Does setting the geometry not cause a repaint?

Reply:
No, the setGeometry doesn't update the panel when the panel's size and layout changes. When the panel moves with its parent and may be placed above or below its parent, the layout can't be updated in real time.